source/common/router/rds_impl.cc

```diff
@@ -195,26 +232,29 @@ Router::RouteConfigProviderSharedPtr RouteConfigProviderManagerImpl::getRdsRoute
   // strong hash instead.
   const std::string manager_identifier = rds.SerializeAsString();
 
-  auto it = route_config_providers_.find(manager_identifier);
-  if (it == route_config_providers_.end()) {
+  RdsRouteConfigSubscriptionSharedPtr subscription;
+
+  auto it = route_config_subscriptions_.find(manager_identifier);
+  if (it == route_config_subscriptions_.end()) {
     // std::make_shared does not work for classes with private constructors. There are ways
     // around it. However, since this is not a performance critical path we err on the side
     // of simplicity.
-    std::shared_ptr<RdsRouteConfigProviderImpl> new_provider{new RdsRouteConfigProviderImpl(
-        rds, manager_identifier, factory_context, stat_prefix, *this)};
-
-    new_provider->registerInitTarget(factory_context.initManager());
+    subscription.reset(new RdsRouteConfigSubscription(rds, manager_identifier, factory_context,
+                                                      stat_prefix, *this));
 
-    route_config_providers_.insert({manager_identifier, new_provider});
+    subscription->registerInitTarget(factory_context.initManager());
 
-    return new_provider;
+    route_config_subscriptions_.insert({manager_identifier, subscription});
+  } else {
+    // Because the RouteConfigProviderManager's weak_ptrs only get cleaned up
+    // in the RdsRouteConfigSubscription destructor, and the single threaded nature
+    // of this code, locking the weak_ptr will not fail.
+    subscription = it->second.lock();
   }
+  ASSERT(subscription);
 
-  // Because the RouteConfigProviderManager's weak_ptrs only get cleaned up
-  // in the RdsRouteConfigProviderImpl destructor, and the single threaded nature
-  // of this code, locking the weak_ptr will not fail.
-  Router::RouteConfigProviderSharedPtr new_provider = it->second.lock();
-  ASSERT(new_provider);
+  Router::RouteConfigProviderSharedPtr new_provider{
```

Comment:
Can this be unique_ptr?

Reply:
It will involving interface changes and make them consistent with StaticRouteConfigProviders, and update `HttpConnectionManager` too.

Reply:
IMHO, it is important to use unique_ptr to show Provider object is not shared, only HttpConnectionManager owns it.